tree.go

```diff
@@ -380,10 +385,15 @@ func (n *InternalNode) InsertStem(stem []byte, values [][]byte, resolver NodeRes
 		// splits.
 		return n.InsertStem(stem, values, resolver)
 	case *LeafNode:
-		n.cowChild(nChild)
 		if equalPaths(child.stem, stem) {
+			// We can't insert anything into a POA leaf node.
+			if child.isPOAStub {
+				return errIsPOAStub
+			}
+			n.cowChild(nChild)
 			return child.insertMultiple(stem, values)
 		}
+		n.cowChild(nChild)
```

Comment:
Interestingly, `(n *InsertNode) Insert(...)` internally calls `InsertStem(...)` (and not `(*LeafNode) Insert(...)`, so we short-circuit here.

Note that unfortunately I had to put `n.cowChild(...)` in two places, to avoid an unwanted side effect of doing that and doing `return errIsPOAStub` later, which would be incorrect. (i.e: unwanted side-effect).

Reply:
My point is that we need to return before calling `n.cowChild(..)`.

If we do `n.cowChild(...)` (saying that this children as changed), and return an error later, it would be incorrect to mark that children as changed.

Not sure if that deserves a comment. It's just something that any API that fails should be aware of (i.e: assuming it will succeed doing stuff, and fail later without reverting the effects).